lib/galaxy/config/__init__.py

```diff
@@ -180,57 +173,142 @@ def _parse_config_file_options(self, defaults, listify_defaults, config_kwargs):
+            setattr(self, var, [os.path.join(self.root, x) for x in paths])
 
 
 class GalaxyAppConfiguration(BaseAppConfiguration):
     deprecated_options = ('database_file', 'track_jobs_in_database')
     default_config_file_name = 'galaxy.yml'
+    # {key: config option, value: deprecated directory name}
+    # If value == first dir in a user path that resolves to key, it will be stripped from the path
+    deprecated_dirs = {'config_dir': 'config', 'data_dir': 'database'}
 
     def __init__(self, **kwargs):
         self._load_schema()  # Load schema from schema definition file
         self._load_config_from_schema()  # Load default propery values from schema
+        self._validate_schema_paths()  # check that paths can be resolved
         self._update_raw_config_from_kwargs(kwargs)  # Overwrite default values passed as kwargs
         self._create_attributes_from_raw_config()  # Create attributes for LOADED properties
+        self._resolve_paths(kwargs)  # Overwrite attributes (not _raw_config) w/resolved paths
         self._process_config(kwargs)  # Finish processing configuration
 
     def _load_schema(self):
         self.schema = AppSchema(GALAXY_CONFIG_SCHEMA_PATH, GALAXY_APP_NAME)
         self.appschema = self.schema.app_schema
 
     def _load_config_from_schema(self):
-        self._raw_config = {}  # keeps track of values provided to the app
-        self.reloadable_options = set()
+        self._raw_config = {}  # keeps track of startup values (kwargs or schema default)
+        self.reloadable_options = set()  # config options we can reload at runtime
+        self._paths_to_resolve = {}  # {config option: referenced config option}
         for key, data in self.appschema.items():
             self._raw_config[key] = data.get('default')
             if data.get('reloadable'):
                 self.reloadable_options.add(key)
+            if data.get('path_resolves_to'):
+                self._paths_to_resolve[key] = data.get('path_resolves_to')
+
+    def _validate_schema_paths(self):
+
+        def check_exists(option, key):
+            if not option:
+                message = "Invalid schema: property '{}' listed as path resolution target " \
+                    "for '{}' does not exist".format(resolves_to, key)
+                raise_error(message)
+
+        def check_type_is_str(option, key):
+            if option.get('type') != 'str':
+                message = "Invalid schema: property '{}' should have type 'str'".format(key)
+                raise_error(message)
+
+        def check_is_dag():
+            def walk(key):
+                if key:
+                    if key in visited:
+                        raise_error('Invalid schema: cycle detected')
+                    visited.add(key)
+                    next = self.appschema[key].get('path_resolves_to')
+                    walk(next)
+
+            for key in self._paths_to_resolve:
+                visited = set()
+                walk(key)
```

Comment:
```suggestion
                    walk(key)
```

Reply:
So, if I got it right, here's the proposed code:
```python
def check_is_dag():
    def walk(key):
        if key in visited:
            raise_error('Invalid schema: cycle detected')
        visited.add(key)
        parent = self._paths_to_resolve[key]
        if parent:
            walk(parent)

    visited = set()
    for key in self._paths_to_resolve:
        if key not in visited:
            walk(key)
```
This won't work. 
Consider this graph: A->B, C->B. Here we have {A, C} in `paths_to_resolve`. When we get to C, we'll then get its parent, which is B, and of course B is in the visited set, so it will look like it found a cycle. (this graph is 2 directories resolved w.r.t. the same parent directory). So we do need to create an empty set for each search path.
Also, this: `parent = self._paths_to_resolve[key]` won't get us B, because B does not need to be resolved (it doesn't have the `path_resolves_to` attribute, so it's not in that collection (there'll be always at least one sink node, otherwise there has to be a cycle).

I'll change `next` to `parent` and will switch the key checks as we discussed.

Reply:
You're right on both counts, thanks for the explanation!
I think the following should work though:

```python
def check_is_dag():
    def visit(key):
        visited.add(key)
        stack.add(key)
        parent = self.appschema[key].get('path_resolves_to')
        if parent:
            if parent not in visited:
                visit(parent)
            elif parent in stack:
                raise_error('Invalid schema: cycle detected')
        stack.remove(key)

    visited = set()
    stack = set()
    for key in self._paths_to_resolve:
        if key not in visited:
            visit(key)
```

This never visits a node twice and has the least recursion calls.

Reply:
I must thank you for this discussion, because through this I think we have an optimal solution.

First, I do see a few issues in the approach above. I wouldn't use a set to represent a stack simply because it's an unordered collection (so it cannot be a LIFO data structure). Also, if we need to do a check like `if foo in bar`, bar should not be a stack (because conceptually, we can only peek or pop from the top of a stack). 

Next, the described cycle detection algorithm is applicable to broader problems. We have a very specific graph: all vertices have at most 1 outgoing edge (because a path cannot resolve to more than one parent), so we never need to backtrack, and so a classic DFS cycle detection algorithm (that looks for back edges) is not needed.

Most importantly, if we use a stack, we don't need recursion at all. The tradeoff is that an iterative algorithm will be usually less intuitive than a recursive algorithm (because we have to maintain our own stack). And I think readability (i.e., developer time) is more important here, given our tiny n (I estimate, unless we are resolving at least 10,000 directories, this won't matter; and in any case we'd profile first.

With that in mind, I was about to make a stack-based iterative solution... Until I realized that for this problem we don't even need a stack (because since a vertex can only have at most 1 outgoing edge, the size of the stack is 0 or 1). Which brings us to a trivial, dirt-simple solution! Here's the  pseudocode:

```
for v in G:
    visited = {}
    while v:
        visited.add(v)
        v = v.next
        if v and v in visited: error (cycle)
```
That's all!

Concrete implementation in next commit. Thank you, again - it was a fun derivation!

Reply:
Is this resolved?